Content.Server/_CD/VoteLink/VoteLinkSystem.cs

```diff
@@ -0,0 +1,210 @@
+using Content.Shared.IdentityManagement;
+using Robust.Server.Player;
+using Robust.Shared.Timing;
+
+namespace Content.Server._CD.VoteLink;
+
+/// <summary>
+///     Handles the station voting system logic
+/// </summary>
+public sealed class VoteLinkSystem : EntitySystem
+{
+    [Dependency] private readonly ChatSystem _chat = default!;
+    [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly IPlayerManager _player = default!;
+    [Dependency] private readonly StationSystem _station = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<StationVoteDatabaseComponent, ComponentInit>(OnDatabaseInit);
+    }
+
+    public void HandleMessage(EntityUid station, EntityUid sender, VoteLinkUiMessageEvent msg)
+    {
+        if (!TryComp<StationVoteDatabaseComponent>(station, out var comp))
+            return;
+
+        switch (msg.Type)
+        {
+            case VoteLinkUiMessageType.Vote:
+                TryCastVote(station, sender, msg.Option);
+                break;
+
+            case VoteLinkUiMessageType.Create:
+                if (msg.VoteData == null)
+                    return;
+
+                // Get the options texts from the VoteData
+                var options = msg.VoteData.Options
+                    .OrderBy(o => o.Option)
+                    .Select(o => o.Text)
+                    .ToList();
+
+                // Get sender's identity for announcement
+                var author = Loc.GetString("comms-console-announcement-unknown-sender"); // recycling
+                var tryGetIdentityShortInfoEvent =
+                    new TryGetIdentityShortInfoEvent(Transform(sender).MapUid ?? sender, sender);
+                RaiseLocalEvent(tryGetIdentityShortInfoEvent);
+                if (!string.IsNullOrEmpty(tryGetIdentityShortInfoEvent.Title))
+                    author = tryGetIdentityShortInfoEvent.Title;
+
+                // Create and dispatch announcement
+                var announcement = Loc.GetString(comp.Message,
+                    ("author", author),
+                    ("question", msg.VoteData.Question));
+
+                _chat.DispatchGlobalAnnouncement(
+                    announcement,
+                    Loc.GetString(comp.Title),
+                    announcementSound: comp.AnnouncementSound);
+
+                // Log it here so we can get the sender
+                _adminLogger.Add(LogType.Action,
+                    LogImpact.Medium,
+                    $"{ToPrettyString(sender)} Started a new vote: '{msg.VoteData.Question}' with options: {string.Join(", ", options)}");
+
+                // Try to start the vote
+                TryStartVote(
+                    station,
+                    msg.VoteData.Question,
+                    options,
+                    msg.VoteData.Duration);
```

Comment:
I think at least the announcement and admin logs should be moved to TryStartVote?

Reply:
I didn't want to pass Actor to the TryStartVote method but it could be a nullable EntityUid